### PR TITLE
Add additional info about when to make strong naming key public

### PR DESCRIPTION
**Pull request by JamesNK**
re: long twitter conversation https://twitter.com/rmhrisk/status/1052315958637993984

Additional detail to explicitly say you shouldn't make a strong name key public if it is used somewhere else that requires it to stay private.

@rmhrisk
@vcsjones
@terrajobst

### CODE REVIEW

**Comment by terrajobst** [comment]
I think we should call it what is because realistically there is only one example:

>  You shouldn't make the strong naming key public if the key was used in the past to grant special permission in partial-trust environments. Otherwise you might compromise existing environments.

/cc @blowdart

**Review by terrajobst — CHANGES_REQUESTED** [review]
We should call out the specific reason to make it less nebulous.

**Comment by blowdart** [comment]
Yea.

Although your wording sucks

"If you have previously strong named closed source components you should use a different strong name key for open source projects, otherwise ALL HELL WILL BREAK LOOSE"

**Comment by mairaw** [comment]
You're missing some hyphens there @blowdart 😄

**Comment by JamesNK** [comment]
How about:

**✔️ CONSIDER** checking in the key used to strong name into your source control system.

> If you have previously strong named closed source components you should use a different strong name key for open source libraries. Using a key that should be kept private might compromise existing environments.

**Comment by terrajobst** [comment]
> Using a key that should be kept private might compromise existing environments.

Still too vague.

**Comment by JamesNK** [comment]
Updated. Is there a way to explain what is meant by partial trust?

**Comment by terrajobst** [comment]
> Is there a way to explain what is meant by partial trust?

Just link to this: https://docs.microsoft.com/en-us/dotnet/framework/misc/using-libraries-from-partially-trusted-code

**Review by terrajobst — APPROVED** [review]
❤️

File: docs/standard/library-guidance/strong-naming.md [hunk 1]
```diff
@@ -46,6 +46,8 @@ You should strong name your open-source .NET libraries. Strong naming an assembl
 **✔️ CONSIDER** checking in the key used to strong name into your source control system.
```
Comment (rpetrusha) [inline docs/standard/library-guidance/strong-naming.md]: nit unrelated to your PR: checking in...into is redundant (should be "checking the key...into...")

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This looks good, @JamesNK. I've left one nit unrelated to your PR, if you'd like to make that change.

**Review by rmhrisk — COMMENTED** [review]
I don’t think “strong naming should be used for security considerations” is appropriate. If the private key is managed appropriately it can be and arguably is better because it identifies a specific thing and not all things from an entity.

**Comment by terrajobst** [comment]
> I don’t think “strong naming should be used for security considerations” is appropriate. If the private key is managed appropriately it can be and arguably is better because it identifies a specific thing and not all things from an entity.

We've [explicit guidance](https://docs.microsoft.com/en-us/dotnet/framework/misc/using-libraries-from-partially-trusted-code) to not use strong naming for security considerations:

> The .NET Framework provides a mechanism for the enforcement of varying levels of trust on different code running in the same application called Code Access Security (CAS). Code Access Security in .NET Framework should not be used as a mechanism for enforcing security boundaries based on code origination or other identity aspects. We are updating our guidance to reflect that Code Access Security and Security-Transparent Code will not be supported as a security boundary with partially trusted code, especially code of unknown origin. We advise against loading and executing code of unknown origins without putting alternative security measures in place.

Also, strong naming shouldn't be used for authenticity as in contrast to AuthentiCode there is no mechanism for revocation. 

So no, the guidance is more than appropriate here.

File: docs/standard/library-guidance/strong-naming.md (hunk 1 above)
Comment (JamesNK) [reply docs/standard/library-guidance/strong-naming.md]: I changed it around a little. I think "check in" in the age of git is anachronistic.

**Comment by rmhrisk** [comment]
> > I don’t think “strong naming should be used for security considerations” is appropriate. If the private key is managed appropriately it can be and arguably is better because it identifies a specific thing and not all things from an entity.
> 
> We've [explicit guidance](https://docs.microsoft.com/en-us/dotnet/framework/misc/using-libraries-from-partially-trusted-code) to not use CAS, which includes using strong naming for security considerations. Also, strong naming shouldn't be used for authenticity as in contrast to Authenticode there is no mechanism for revocation.
> 
> So no, the guidance is more than appropriate here.

The statement in question is about the strong names and not CAS, I recommend:

```
You shouldn't make the strong naming private key public if it has been used in the past to give special permissions in [partial-trust scenarios](/dotnet/framework/misc/using-libraries-from-partially-trusted-code). Otherwise, you might compromise existing environments.

[!IMPORTANT]
When a the identity of the publisher of the code is desired, [Authenticode](/windows-hardware/drivers/install/authenticode) and [NuGet Package Signing](/nuget/create-packages/sign-a-package) are recommended. CAS should not be used as a security mitigation.
```

**Comment by terrajobst** [comment]
> I recommend:

That wording looks good to me.

@JamesNK, mind updating it?

**Comment by JamesNK** [comment]
🆙 📅

**Review by rpetrusha — APPROVED** [review]

**Comment by rpetrusha** [comment]
This looks good. Shall I merge it, @JamesNK?

**Comment by JamesNK** [comment]
Go ahead

**Pull request merged**